protobuf 3.18.0 broken for Windows when building with MinGW-w64

protobuf 3.18.0 fails to build on Windows with MinGW-w64 GCC with the following errors:
```
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:108:28: error: 'F_GETFL' was not declared in this scope
  108 |   int flags = fcntl(file_, F_GETFL);
      |                            ^~~~~~~
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:108:15: error: 'fcntl' was not declared in this scope
  108 |   int flags = fcntl(file_, F_GETFL);
      |               ^~~~~
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:109:13: error: 'O_NONBLOCK' was not declared in this scope
  109 |   flags &= ~O_NONBLOCK;
      |             ^~~~~~~~~~
R:/winlibs64-10.3.0/protobuf-3.18.0/src/google/protobuf/io/zero_copy_stream_impl.cc:110:16: error: 'F_SETFL' was not declared in this scope
  110 |   fcntl(file_, F_SETFL, flags);
      |                ^~~~~~~
```

The solution is to replace `_MSC_VER` with `_WIN32` in `src/google/protobuf/io/zero_copy_stream_impl.cc`.

Apparently a false assumption was made `_MSC_VER` is always set on the Windows platform, but that is only the case when building with MSVC. On MSVC and MinGW gcc and clang `_WIN32` is always defined however.

--- a/ports/protobuf/fix-mingw-build.patch
+++ b/ports/protobuf/fix-mingw-build.patch
@@ -1,0 +1,20 @@
+--- a/src/google/protobuf/io/zero_copy_stream_impl.cc	2021-09-15 00:48:28.000000000 +0800
++++ b/src/google/protobuf/io/zero_copy_stream_impl.cc	2022-01-05 09:42:35.632457100 +0800
+@@ -32,7 +32,7 @@
+ //  Based on original Protocol Buffers design by
+ //  Sanjay Ghemawat, Jeff Dean, and others.
+ 
+-#ifndef _MSC_VER
++#ifndef _WIN32
+ #include <fcntl.h>
+ #include <sys/stat.h>
+ #include <sys/types.h>
+@@ -104,7 +104,7 @@
+       is_closed_(false),
+       errno_(0),
+       previous_seek_failed_(false) {
+-#ifndef _MSC_VER
++#ifndef _WIN32
+   int flags = fcntl(file_, F_GETFL);
+   flags &= ~O_NONBLOCK;
+   fcntl(file_, F_SETFL, flags);

--- a/ports/protobuf/fix-mingw-build.patch
+++ b/ports/protobuf/fix-mingw-build.patch
@@ -1,14 +1,5 @@
 --- a/src/google/protobuf/io/zero_copy_stream_impl.cc	2021-09-15 00:48:28.000000000 +0800
-+++ b/src/google/protobuf/io/zero_copy_stream_impl.cc	2022-01-05 09:42:35.632457100 +0800
-@@ -32,7 +32,7 @@
- //  Based on original Protocol Buffers design by
- //  Sanjay Ghemawat, Jeff Dean, and others.
- 
--#ifndef _MSC_VER
-+#ifndef _WIN32
- #include <fcntl.h>
- #include <sys/stat.h>
- #include <sys/types.h>
++++ b/src/google/protobuf/io/zero_copy_stream_impl.cc	2022-01-06 15:19:12.657413400 +0800
 @@ -104,7 +104,7 @@
        is_closed_(false),
        errno_(0),

--- a/ports/protobuf/portfile.cmake
+++ b/ports/protobuf/portfile.cmake
@@ -8,6 +8,7 @@ vcpkg_from_github(
         fix-static-build.patch
         fix-default-proto-file-path.patch
         fix-uwp-build.patch
+        fix-mingw-build.patch
 )
 
 string(COMPARE EQUAL "${TARGET_TRIPLET}" "${HOST_TRIPLET}" protobuf_BUILD_PROTOC_BINARIES)

--- a/ports/protobuf/vcpkg.json
+++ b/ports/protobuf/vcpkg.json
@@ -1,6 +1,7 @@
 {
   "name": "protobuf",
   "version-semver": "3.18.0",
+  "port-version": 1,
   "description": "Protocol Buffers - Google's data interchange format",
   "homepage": "https://github.com/protocolbuffers/protobuf",
   "dependencies": [

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -5442,7 +5442,7 @@
     },
     "protobuf": {
       "baseline": "3.18.0",
-      "port-version": 0
+      "port-version": 1
     },
     "protobuf-c": {
       "baseline": "1.4.0",

--- a/versions/p-/protobuf.json
+++ b/versions/p-/protobuf.json
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "166c456962720a9382505f3d0cdf6b2fc986c167",
+      "git-tree": "5f6aa04d94e184b89c1887946fb4684bc51c4611",
       "version-semver": "3.18.0",
       "port-version": 1
     },

--- a/versions/p-/protobuf.json
+++ b/versions/p-/protobuf.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "166c456962720a9382505f3d0cdf6b2fc986c167",
+      "version-semver": "3.18.0",
+      "port-version": 1
+    },
+    {
       "git-tree": "19c2bf45c235e6126161bae36aa7ff7e1fcda479",
       "version-semver": "3.18.0",
       "port-version": 0

